external-crates/move/crates/move-compiler/src/diagnostics/warning_filters.rs

```diff
@@ -67,24 +69,42 @@ pub struct WarningFiltersScope(WarningFiltersScope_);
 enum WarningFiltersScope_ {
     /// Unsafe and should be used only for internal purposes, such as ide annotations
     Empty,
-    Static(&'static WarningFilters),
+    /// The top-level warning filters given to the compiler instance. They are leaked as they will
+    /// be needed for the lifetime of the compiler instance.
+    Static(&'static WarningFiltersBuilder),
+    /// A user-defined warning filter scope, with a reference to the previous scope
     Node(Arc<WarningFiltersScopeNode>),
 }
 
 #[derive(PartialEq, Eq, Clone, Debug)]
 struct WarningFiltersScopeNode {
+    /// The warning filters for this scope
     filters: WarningFilters,
+    /// The previous scope
     prev: WarningFiltersScope_,
 }
 
-#[derive(PartialEq, Eq, Clone, Debug)]
+#[derive(Debug, Clone)]
+/// An intern table for warning filters. The `Pin` is likely unnecessary given the usage of
+/// `WarningFiltersTable`, but it is added to ensure the underlying `Box` is not moved.
+/// Safety: This table should not be dropped as long as any `WarningFilters` are alive
+pub struct WarningFiltersTable(HashSet<Pin<Box<WarningFiltersBuilder>>>);
```

Comment:
Open to suggestions about my intern table here. Maybe there is something slightly more clever we can do?

Reply:
I have a pretty drastic change suggestion:

What if we number the well-known filters? And then we can use a u64 (or, if we ever get more than 64, go up in size to a u128/u256/bitset). `All` can be `u64::MAX`, `none` is `0`, and we just do mask checks. Then interning is just writing down that specific u64, and checking is a quick walk down a chain of those.

Reply:
Also I agree with the comment, I don't think you need `Pin` here. Try it without and see if it breaks?

Reply:
I removed the pin and all seems to be well. 

I looked into the bitset approach, but I got stuck with external filters. The warning filters being allowed to come from outside sources for warnings not defined by the compiler makes this all a bit tricky